mitiq/lre/multivariate_scaling/layerwise_folding.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (C) Unitary Fund
+#
+# This source code is licensed under the GPL license (v3) found in the
+# LICENSE file in the root directory of this source tree.
+
+"""Functions for layerwise folding of input circuits to allow for multivariate
+extrapolation."""
+
+import itertools
+from copy import deepcopy
+
+import cirq
+import numpy as np
+
+from mitiq.utils import _pop_measurements
+from mitiq.zne.scaling.folding import _check_foldable
+
+
+def _get_num_layers_without_measurements(input_circuit: cirq.Circuit) -> int:
+    """Checks if the circuit has non-terminal measurements and returns the
+    number of layers in the input circuit without the terminal measurements.
+
+        Args:
+            input_circuit: Checks whether this circuit is able to be folded.
+
+        Raises:
+            UnfoldableCircuitError:
+                * If the circuit has intermediate measurements.
+                * If the circuit has non-unitary channels which are not
+                terminal measurements.
+
+        Returns:
+            num_layers: the number of layers in the input circuit without the
+            terminal measurements.
+    """
+    circuit = deepcopy(input_circuit)
+    _check_foldable(circuit)
+    _pop_measurements(circuit)
+    num_layers = len(circuit)
+    return num_layers
+
+
+def _get_chunks(
+    input_circuit: cirq.Circuit, num_chunks: int
+) -> list[cirq.Circuit]:
+    """Splits a circuit into approximately equal chunks.
+
+    Adapted from:
+    https://stackoverflow.com/questions/2130016/splitting-a-list-into-n-parts-of-approximately-equal-length
+
+        Args:
+            input_circuit: Circuit of interest.
+            num_chunks: Number of desired approximately equal chunks
+            * when num_chunks == num_layers, the circuit gets split into each
+                layer in the input circuit
+            * when num_chunks == 1, the entire circuit is chunked into 1 layer
+
+
+        Returns:
+            split_circuit: Circuit of interest split into approximately equal
+            chunks
+    """
+    num_layers = _get_num_layers_without_measurements(input_circuit)
+    if num_chunks == 0:
+        raise ValueError("The number of chunks should be >= to 1.")
+
+    if num_chunks > num_layers:
+        raise ValueError(
+            "Number of chunks > the number of layers in the circuit."
+        )
+
+    k, m = divmod(num_layers, num_chunks)
+    split_circuit = (
+        input_circuit[i * k + min(i, m) : (i + 1) * k + min(i + 1, m)]
+        for i in range(num_chunks)
+    )
+    return list(split_circuit)
+
+
+def _get_scale_factor_vectors(
+    input_circuit: cirq.Circuit,
+    degree: int,
+    fold_multiplier: int,
+    num_chunks: int = 1,
```

Comment:
It would be better to use the default value as `None` for the parameter `num_chunks`. Making changes to this currently leads to the circuit getting chunked for some reason. Figure this out; make changes.